File: leetcode/Unique_Paths.java
import java.util.*;

public class Solution {
    public int uniquePaths(int m, int n) {
        int a = m - 1 + n - 1;
        int b = Math.max(m - 1, n - 1);
        int c = a - b;
        long ret = 1;
        while (a > c && b > 1) {
            ret *= a;
            a--;
            if (ret % b == 0) {
                ret /= b;
                b--;
            }
        }
        while (a > c) {
            ret *= a;
            a--;
        }
        while (b > 1) {
            ret /= b;
            b--;
        }
        return (int)ret;
    }
}

File: leetcode/Unique_Paths.cpp
class Solution {
    int calculate(int n, int k) {  // Select k items from 1..n
        long long res = 1LL;
        int b = k;
        for (int a = n; a >= n - k + 1; --a) {
            res *= a;
            while (b > 1 && res % b == 0) {
                res /= b;
                b -= 1;
            }
        }
        return (int)res;
    }
public:
    int uniquePaths(int m, int n) {
        if (m == 0 || n == 0)
            return 0;
        return calculate(m + n - 2, n - 1);
    }
};
